File: external/boost_1.55.0/libs/python/test/properties.py
# Copyright David Abrahams 2004. Distributed under the Boost
# Software License, Version 1.0. (See accompanying
# file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
"""
This is test module for properties.

>>> r = properties.ret_type()
>>> r.i = 22.5
>>> r.i
22.5
>>> c = properties.crash_me()
>>> c.i.i
42.5

>>> X = properties.X

>>> x1 = X(1)

value read only
>>> x1.value_r
1

value read - write
>>> x1.value_rw
1

value direct access
>>> x1.value_direct
1

class instance count read - only
>>> X.instance_count
1

class instance count direct
>>> X.instance_count_direct
1

class instance count injected
>>> X.instance_count_injected
1

class instance count from object
>>> x1.instance_count
1

class instance count from object
>>> x1.instance_count_direct
1

class instance count from object: 
>>> x1.instance_count_injected
1

as expected you can't assign new value to read only property
>>> x1.value_r = 2
Traceback (most recent call last):
  File "properties.py", line 49, in ?
    x1.value_r = 2
AttributeError: can't set attribute

setting value_rw to 2. value_direct: 
>>> x1.value_rw = 2
>>> x1.value_rw
2

setting value_direct to 3. value_direct: 
>>> x1.value_direct = 3
>>> x1.value_direct
3

>>> assert x1.value_r == 3

>>> x2 = X(2)

after creating second intstance of X instances count is 2
>>> x2.instance_count
2

>>> del x2
>>> assert x1.instance_count == 1

>>> assert properties.X.value_r_ds.__doc__ == "value_r_ds is read-only"

>>> assert properties.X.value_rw_ds.__doc__ == "value_rw_ds is read-write"

"""

#import sys; sys.path.append(r'P:\Actimize4.0\smart_const\py_smart_const___Win32_Debug')
import properties_ext as properties


def run(args = None):
    import sys
    import doctest

    if args is not None:
        sys.argv = args
    return doctest.testmod(sys.modules.get(__name__))
    
if __name__ == '__main__':
    print "running..."
    import sys
    status = run()[0]
    if (status == 0): print "Done."
    sys.exit(status)

File: external/boost_1.55.0/libs/python/test/properties.cpp
// Copyright David Abrahams 2004. Distributed under the Boost
// Software License, Version 1.0. (See accompanying
// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
#include <boost/python.hpp>

using namespace boost::python;

namespace test {

// Hmm.  return_internal_reference<>() wants to wrap a real class.
class ret_type
{
 public:
    ret_type() : i(42.5) {}
    double i;
};

class crash_me
{
    private:
        ret_type i;
    public:
        ret_type& get_i() { return i; }
};

}

struct X
{
    X( int value ) : m_value( value )
    { ++s_count; }
    
    X( const X &other ) : m_value( other.m_value )
    { ++s_count; }
    
    ~X() 
    { --s_count; }
        
    int get_value() const
    { return m_value; }
    
    void set_value(int new_value)
    { m_value = new_value; }
    
    static int get_instance_count()
    { return s_count; }
        
    int m_value;

    static int s_count;
};

int X::s_count = 0;

int get_X_instance_count()
{ return X::get_instance_count(); }



BOOST_PYTHON_MODULE(properties_ext)
{
    typedef return_value_policy<return_by_value> return_by_value_t;
    typedef return_internal_reference<> return_by_internal_reference_t;
    class_<X>("X", init<int>() )
        //defining read only property
        .add_property( "value_r", &X::get_value )
        .add_property( "value_r_ds", &X::get_value, "value_r_ds is read-only")
        //defining read \ write property 
        .add_property( "value_rw", &X::get_value, &X::set_value )
        .add_property( "value_rw_ds", &X::get_value, &X::set_value, 
            "value_rw_ds is read-write")
        //defining read \ write property using make_getter and make_setter
        .add_property( "value_direct", 
                        make_getter( &X::m_value, return_by_value_t() ),
                        make_setter( &X::m_value, return_by_internal_reference_t() ) )
        //defining read only property for static member
        .add_static_property( "instance_count", &X::get_instance_count )
        //defining read \ write property for static member using make_getter and make_setter
        .add_static_property( "instance_count_direct", 
                              make_getter( &X::s_count, return_by_value_t() ), 
                              make_setter( &X::s_count, return_by_internal_reference_t() ) )
        //defining class property using a global function
        .add_static_property( "instance_count_injected", &get_X_instance_count );

    
    class_< test::ret_type>( "ret_type")
        .add_property( "i", &test::ret_type::i, &test::ret_type::i)
        ;
    
    class_< test::crash_me> crash_me_wrapper( "crash_me");
    
    crash_me_wrapper
        .def( "get_i", &test::crash_me::get_i , return_internal_reference<>())
        ;
    
    crash_me_wrapper.add_property( "i", crash_me_wrapper.attr("get_i"));

}

#include "module_tail.cpp"
